### ISSUE TITLE
file.find unicode errors in 2016.11.6

### ISSUE THREAD

**Issue by mbochenk** [saltstack/salt#42292]
### Description of Issue/Question
file.find fails if the search path have directories with unicode characters: `UnicodeDecodeError: 'ascii' codec can't decode byte`
### Setup
Very similar to #40578 - creating a new issue per @terminalmage request.
CentOS6/7. Note that this was not an issue on 2016.11.1 and I found it in 2016.11.5 and 2016.11.6 - other versions could also be affected.

Searching for regex pattern in a directory on a mounted fs. As mount is not directly controlled, it may have ascii characters in some paths. 

### Steps to Reproduce Issue
Original comment https://github.com/saltstack/salt/issues/40578#issuecomment-312297344
Call: `{% set copy_path = salt.file.find(...)|sort|last %}` with parameters fails. Log from CentOS6:
```
10:56:03			Traceback (most recent call last):
10:56:03			  File "/usr/lib/python2.6/site-packages/salt/utils/templates.py", line 368, in render_jinja_tmpl
10:56:03			    output = template.render(**decoded_context)
10:56:03			  File "/usr/lib/python2.6/site-packages/jinja2/environment.py", line 1008, in render
10:56:03			    return self.environment.handle_exception(exc_info, True)
10:56:03			  File "/usr/lib/python2.6/site-packages/jinja2/environment.py", line 780, in handle_exception
10:56:03			    reraise(exc_type, exc_value, tb)
10:56:03			  File "<template>", line 10, in top-level template code
10:56:03			  File "...sls", line 95, in top-level template code
10:56:03			    {% set copy_path = salt.file.find(...)|sort|last %}
10:56:03			  File "/usr/lib/python2.6/site-packages/salt/modules/file.py", line 880, in find
10:56:03			    ret = [item for i in [finder.find(p) for p in glob.glob(os.path.expanduser(path))] for item in i]
10:56:03			  File "/usr/lib/python2.6/site-packages/salt/utils/find.py", line 642, in find
10:56:03			    for dirpath, dirs, files in os.walk(path):
10:56:03			  File "/usr/lib64/python2.6/os.py", line 284, in walk
10:56:03			    if isdir(join(top, name)):
10:56:03			  File "/usr/lib64/python2.6/posixpath.py", line 70, in join
10:56:03			    path += '/' + b
10:56:03			UnicodeDecodeError: 'ascii' codec can't decode byte 0xef in position 7: ordinal not in range(128)
```

### Versions Report
```
Salt Version:
           Salt: 2016.11.6

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.4.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: 1.3.8
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: 1.4.3
       M2Crypto: 0.20.2
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.6.6 (r266:84292, Aug 18 2016, 15:13:37)
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 14.5.0
           RAET: Not Installed
          smmap: 0.8.1
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: centos 6.9 Final
        machine: x86_64
        release: 2.6.32-696.6.3.el6.x86_64
         system: Linux
        version: CentOS 6.9 Final
```

**Comment by Ch3LL** [comment]
@mbochenk i'm having a hard time replicating this. Here is my test case:

```
{% set copy_path = salt.file.find('/home/', type='f')|sort|last %}

echo {{ copy_path }}:
  cmd.run
```

```
[root@95ed2bbd6345 /]# cat /home/blah 
名
```

```
[root@95ed2bbd6345 /]# salt-call --local state.sls test                                                                    [0/72028]
local:                                          
----------                                         
          ID: echo /home/blah                         
    Function: cmd.run                           
      Result: True                              
     Comment: Command "echo /home/blah" run        
     Started: 17:13:29.729420                         
    Duration: 121.13 ms 
     Changes:           
              ----------
              pid:      
                  370   
              retcode:  
                  0     
              stderr:                           
              stdout:                              
                  /home/blah                          
                        
Summary for local       
------------            
Succeeded: 1 (changed=1)
Failed:    0            
------------                                    
Total states run:     1                            
Total run time: 121.130 ms 
```

And as you can see it runs it correctly. Can you see anything wrong with my test case? or any other information you can supply to help replicate?

**Comment by mbochenk** [comment]
@Ch3LL I'd need to set up environment to test it as I worked around it by removing unicode chars from path names to give you more details.
From what I have on me now: one of the characters in question that turned out to be difficult to track was `ef80a8` or `0xF028`. From what I understand while not being a 'character' per se the code is still valid and should processed.

**Comment by damon-atkins** [comment]
Ping @Ch3LL

**Comment by untra** [comment]
👋 Heya folks encountering the `UnicodeDecodeError: 'ascii' codec can't decode byte 0xef in position 0: ordinal not in range(128)`

We have encountered this issue when using the [salt file.contains module](https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.file.html#salt.modules.file.contains) to validate settings in configuration files we thought were appropriately ascii / utf-8 encoded.

The problem turned out to be a [utf-8 byte-order-marker](https://en.wikipedia.org/wiki/Byte_order_mark#UTF-8) at the start of the file, which is a three byte `EF BB BF` sequence which is clearly outside the ascii range. This invisible byte triplet is hidden in files viewed in windows notepad, and can cause headaches with salt blowing up unable to handle a _seemingly reasonable_ ascii / utf-8 encoded files.

I can open a new issue for this (title: salt file.contains module will throw when attempting to parse utf-8 encoded files with BOM) , but considering utf-8 BOMs are strongly discouraged in the first place and are a artifact of legacy windows file encoding, I will leave this at the discretion of the salt developers to address.

**Comment by Ch3LL** [comment]
@untra go ahead and open a new issue with more details like the salt version, state file, the file causing the issue and so forth. We made a lot of improvements with unicode handling in 2019.2, so would be curious to see what version and what the use case is.

**Comment by untra** [comment]
@Ch3LL thanks for the reply!

we are using 2018.3 windows minions, although we will give 2019.2 a try. Honestly we removed the BOM from the (admittedly very old) file and everything started working again. I'm not sure how prolific utf-8 BOMs are in the wild, but I sure hope they are a fluke 🤞  .

I opened issue #54357 to elaborate and get it in your ticketing system. Our team has moved on past this, but I'll help see the new issue through 👍
